In [1]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [2]:
import numpy as np
import pandas as pd
import os

csvPath = '/content/drive/MyDrive/구글클라우드인공지능개발자과정/공공데이터포털/제주도태양광/제주특별자치도_태양광발전소현황_20220630.csv'
df = pd.read_csv(csvPath, encoding = 'cp949')

In [3]:
df.head()

행정시  읍면동        허가일자            상호  설비용량(KW)    상태       사업개시일     데이터기준일자
0  제주시  구좌읍  1998-06-10  행원풍력발전단지(1차)    3480.0  사업개시  2000-04-05  2022-06-30
1  제주시  회천동  2002-12-09     파낙스에너지(주)    1000.0  사업개시  2002-12-18  2022-06-30
2  제주시  한경면  2003-04-14      한경풍력 1단계    6000.0  사업개시  2004-02-28  2022-06-30
3  제주시  한경면  2005-04-04      신창풍력발전단지    1700.0  사업개시  2006-03-03  2022-06-30
4  제주시  구좌읍  2005-08-30     제주월정풍력발전소    1500.0  사업개시  2006-07-20  2022-06-30

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2148 entries, 0 to 2147
Data columns (total 8 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   행정시       2148 non-null   object 
 1   읍면동       2148 non-null   object 
 2   허가일자      2148 non-null   object 
 3   상호        2148 non-null   object 
 4   설비용량(KW)  2148 non-null   float64
 5   상태        2148 non-null   object 
 6   사업개시일     1571 non-null   object 
 7   데이터기준일자   2148 non-null   object 
dtypes: float64(1), object(7)
memory usage: 134.4+ KB


## df에서 사업개시일 nan값인 데이터 없애기

In [5]:
df['사업개시일'].notnull().sum()

1571

In [6]:
df['상태'].value_counts()

사업개시    1568
인허가      480
공사진행     100
Name: 상태, dtype: int64

오류 3개가 있다. 이걸 처리한 뒤에 .dropna() 처리할 것.

In [7]:
test_df = df[df['사업개시일'].notnull()]

In [8]:
test_df[test_df['상태'] != '사업개시']

행정시  읍면동        허가일자         상호  설비용량(KW)    상태       사업개시일     데이터기준일자
6     제주시  한경면  2006-08-11  탐라해상풍력발전소   30000.0  공사진행  2017-09-16  2022-06-30
29   서귀포시  성산읍  2008-02-28   성산풍력 1단계   12000.0  공사진행  2009-03-30  2022-06-30
583   제주시  한경면  2016-11-16     성용유화발전     750.0   인허가  2016-11-16  2022-06-30

탐라해상은 운영 중, 성산풍력은 이미 준공 완료한 거 보니 운영 중인듯, 성용유화는 2019년 12월 당시에도 추진 중.

앞 2개는 상태를 사업개시로, 뒤 1개는 사업개시일을 값 삭제.

In [9]:
correctList = list(test_df.index[(test_df['상태'] != '사업개시')])

In [10]:
correctList

[6, 29, 583]

In [11]:
for i in correctList:
  if i == correctList[2]:
    df.loc[i, '사업개시일'] = np.nan
  else:
    df.loc[i, '상태'] = '사업개시'

In [12]:
df.iloc[correctList]

행정시  읍면동        허가일자         상호  설비용량(KW)    상태       사업개시일     데이터기준일자
6     제주시  한경면  2006-08-11  탐라해상풍력발전소   30000.0  사업개시  2017-09-16  2022-06-30
29   서귀포시  성산읍  2008-02-28   성산풍력 1단계   12000.0  사업개시  2009-03-30  2022-06-30
583   제주시  한경면  2016-11-16     성용유화발전     750.0   인허가         NaN  2022-06-30

In [13]:
df.dropna(inplace = True)

In [14]:
df.reset_index(drop=True, inplace=True)

In [15]:
df

행정시  읍면동        허가일자            상호  설비용량(KW)    상태       사업개시일  \
0      제주시  구좌읍  1998-06-10  행원풍력발전단지(1차)   3480.00  사업개시  2000-04-05   
1      제주시  회천동  2002-12-09     파낙스에너지(주)   1000.00  사업개시  2002-12-18   
2      제주시  한경면  2003-04-14      한경풍력 1단계   6000.00  사업개시  2004-02-28   
3      제주시  한경면  2005-04-04      신창풍력발전단지   1700.00  사업개시  2006-03-03   
4      제주시  구좌읍  2005-08-30     제주월정풍력발전소   1500.00  사업개시  2006-07-20   
...    ...  ...         ...           ...       ...   ...         ...   
1565  서귀포시  안덕면  2021-04-19      썬 태양광발전소    182.00  사업개시  2022-01-03   
1566  서귀포시  안덕면  2021-04-19     청춘 태양광발전소    662.48  사업개시  2022-01-03   
1567  서귀포시  표선면  2021-04-21   제주덕암 태양광발전소     99.18  사업개시  2022-01-03   
1568  서귀포시  안덕면  2021-08-03    태양7호태양광발전소    198.88  사업개시  2021-12-10   
1569   제주시  구좌읍  2021-09-15      흥산태양광발전소    999.63  사업개시  2022-04-25   

         데이터기준일자  
0     2022-06-30  
1     2022-06-30  
2     2022-06-30  
3     2022-06-30  
4     2022-06-30  
...          ...  
1565  2022-06-30  
1566  2022-06-30  
1567  2022-06-30  
1568  2022-06-30  
1569  2022-06-30  

[1570 rows x 8 columns]

## 상호명으로 거르기(생략)

## df2 파일로 거르기

In [16]:
csvPath2 = '/content/drive/MyDrive/구글클라우드인공지능개발자과정/공공데이터포털/제주도태양광/제주특별자치도_신재생에너지발전시설현황_20200120.csv'
df2 = pd.read_csv(csvPath2, encoding = 'cp949')

In [17]:
df2.head()

허가일자                        상호  설비용량(KW) 설치행정시 설치지역 원동력종류  \
0    2018.4.4           (유)나월(나월태양광발전소)     496.8   제주시  애월읍   태양광   
1   2014.10.7            (유)메가솔라 태양광발전소    1800.0   제주시  구좌읍   태양광   
2   2014.5.15  (유)무한에코에너지(서민행복8호태양광발전소)     516.0  서귀포시  남원읍   태양광   
3   2014.12.9             (유)용흥마을태양광발전소      99.0  서귀포시  강정동   태양광   
4  2014.10.14                 (유)제주 늘해랑    2198.0  서귀포시  성산읍   태양광   

        사업개시일  구분    데이터기준일자  
0   2019.4.26  육상  2020.1.20  
1    2015.7.1  없음  2020.1.20  
2  2014.10.16  없음  2020.1.20  
3  2015.10.29  없음  2020.1.20  
4   2019.1.31  없음  2020.1.20

In [18]:
df2[df2['원동력종류'] == '태양광'].head()

허가일자                        상호  설비용량(KW) 설치행정시 설치지역 원동력종류  \
0    2018.4.4           (유)나월(나월태양광발전소)     496.8   제주시  애월읍   태양광   
1   2014.10.7            (유)메가솔라 태양광발전소    1800.0   제주시  구좌읍   태양광   
2   2014.5.15  (유)무한에코에너지(서민행복8호태양광발전소)     516.0  서귀포시  남원읍   태양광   
3   2014.12.9             (유)용흥마을태양광발전소      99.0  서귀포시  강정동   태양광   
4  2014.10.14                 (유)제주 늘해랑    2198.0  서귀포시  성산읍   태양광   

        사업개시일  구분    데이터기준일자  
0   2019.4.26  육상  2020.1.20  
1    2015.7.1  없음  2020.1.20  
2  2014.10.16  없음  2020.1.20  
3  2015.10.29  없음  2020.1.20  
4   2019.1.31  없음  2020.1.20

In [19]:
df.shape, df2[df2['원동력종류'] == '태양광'].shape

((1570, 8), (811, 9))

In [20]:
solarList = list(df2[df2['원동력종류'] == '태양광']['상호'])

여기 데이터 기준일자가 20년 1월 20일이므로 우선 df에서도 이 날 이후 사업개시한 행은 제거해줘야 함.

In [21]:
df[df['사업개시일'] <= '2020-01-20']

행정시  읍면동        허가일자               상호  설비용량(KW)    상태       사업개시일  \
0      제주시  구좌읍  1998-06-10     행원풍력발전단지(1차)   3480.00  사업개시  2000-04-05   
1      제주시  회천동  2002-12-09        파낙스에너지(주)   1000.00  사업개시  2002-12-18   
2      제주시  한경면  2003-04-14         한경풍력 1단계   6000.00  사업개시  2004-02-28   
3      제주시  한경면  2005-04-04         신창풍력발전단지   1700.00  사업개시  2006-03-03   
4      제주시  구좌읍  2005-08-30        제주월정풍력발전소   1500.00  사업개시  2006-07-20   
...    ...  ...         ...              ...       ...   ...         ...   
1280   제주시  애월읍  2019-08-16        근민 태양광발전소     99.90  사업개시  2019-12-02   
1282   제주시  애월읍  2019-08-16      해맑은2 태양광발전소     97.11  사업개시  2019-12-10   
1287   제주시  한림읍  2019-08-19         양천태양광발전소    291.60  사업개시  2019-11-30   
1296   제주시  애월읍  2019-08-29       광덕태양광1호발전소    382.50  사업개시  2020-01-14   
1298  서귀포시  안덕면  2019-09-03  제이씨지솔라3호 태양광발전소    310.80  사업개시  2020-01-10   

         데이터기준일자  
0     2022-06-30  
1     2022-06-30  
2     2022-06-30  
3     2022-06-30  
4     2022-06-30  
...          ...  
1280  2022-06-30  
1282  2022-06-30  
1287  2022-06-30  
1296  2022-06-30  
1298  2022-06-30  

[914 rows x 8 columns]

In [22]:
df_before200120 = df[df['사업개시일'] <= '2020-01-20']

## solarList에는 있는데 df에는 없는 회사 찾아야 함

In [23]:
len(solarList)

811

In [24]:
count = 0
for i in solarList:
  if i in list(df['상호']):
    count += 1
count

770

41개 회사가 있음.

In [25]:
nonexistList = []
for i in solarList:
  if i not in list(df['상호']):
    nonexistList.append(i)
nonexistList

['(주)가시리태양광',
 '(주)남정수상태양광((주)남정수상태양광(상모1호기))',
 '(주)삼우태양광발전',
 '(주)상은태양광',
 '(주)제주햇빛발전소(제주햇빛발전소)',
 '(주)표선태양광',
 '(주)한솔태양광',
 '17-001 신성여중학교태양광',
 'AJ태양광발전소',
 'JCG에너지2호태양광발전소',
 'JCG에너지3호태양광발전소',
 '건희태양광발전소',
 '길 태양광발전소',
 '다사랑태양광발전소',
 '다온태양광발전소',
 '동신이엔텍태양광발전소제주',
 '동캐1호태양광발전소',
 '두못개태양광발전소',
 '명승 태양광발전소',
 '민솔라에너지(금빛태양광발전소 양수)',
 '반석 태양광발전소',
 '봉개동주민대책위원회태양광발전소',
 '서영1호태양광발전소',
 '서영2호태양광발전소',
 '선우태양광발전소',
 '월림햇빛발전소',
 '이호동 오도마을회',
 '조운1호태양광발전소',
 '조운2호태양광발전소',
 '주식회사 세왕',
 '진희1호태양광발전소',
 '창희2호태양광발전소',
 '천혜태양광발전소',
 '케이원에너지 1호 태양광발전소',
 '케이원에너지 2호 태양광발전소',
 '태성2호태양광발전소',
 '한동1호태양광발전소',
 '햇볕1호태양광발전소',
 '행원2018 1호태양광발전소',
 '현원2호태양광발전소',
 '혜경태양광발전소']

In [26]:
existList = []
for i in solarList:
  if i in list(df['상호']):
    existList.append(i)
len(existList)

770

최종적으로 exisList 가지고 df에서 걸러내면 됨!

In [27]:
df[df['상호'] == existList[0]]

행정시  읍면동        허가일자               상호  설비용량(KW)    상태       사업개시일  \
804  제주시  애월읍  2018-04-04  (유)나월(나월태양광발전소)     496.8  사업개시  2019-04-26   

        데이터기준일자  
804  2022-06-30

In [28]:
df.index[df['상호'] == existList[0]][0]

804

In [29]:
indexList = []

for i in existList:
  indexList.append(df.index[df['상호'] == i][0])

In [30]:
indexList

[804,
 329,
 265,
 417,
 343,
 22,
 798,
 459,
 207,
 726,
 72,
 334,
 295,
 294,
 229,
 234,
 218,
 280,
 261,
 202,
 225,
 137,
 137,
 71,
 11,
 228,
 1215,
 313,
 258,
 257,
 233,
 590,
 32,
 124,
 246,
 232,
 453,
 80,
 77,
 127,
 171,
 247,
 256,
 190,
 238,
 347,
 355,
 354,
 264,
 711,
 74,
 352,
 727,
 572,
 363,
 606,
 321,
 274,
 306,
 628,
 533,
 647,
 195,
 824,
 131,
 753,
 623,
 826,
 891,
 364,
 377,
 378,
 376,
 637,
 708,
 1153,
 784,
 989,
 988,
 936,
 158,
 582,
 521,
 602,
 26,
 16,
 16,
 1246,
 46,
 157,
 946,
 1170,
 494,
 248,
 326,
 499,
 388,
 577,
 575,
 576,
 861,
 340,
 641,
 457,
 269,
 1280,
 528,
 529,
 214,
 215,
 816,
 42,
 589,
 573,
 553,
 619,
 618,
 1221,
 519,
 216,
 485,
 213,
 212,
 182,
 741,
 506,
 283,
 489,
 610,
 307,
 67,
 252,
 391,
 906,
 270,
 254,
 78,
 112,
 534,
 1274,
 1156,
 93,
 292,
 429,
 717,
 75,
 539,
 25,
 369,
 415,
 400,
 448,
 84,
 792,
 199,
 638,
 412,
 69,
 940,
 800,
 470,
 325,
 149,
 419,
 893,
 285,
 342,
 397,
 180

In [31]:
df.loc[indexList]

행정시  읍면동        허가일자                        상호  설비용량(KW)    상태  \
804    제주시  애월읍  2018-04-04           (유)나월(나월태양광발전소)    496.80  사업개시   
329    제주시  구좌읍  2014-10-07            (유)메가솔라 태양광발전소   1800.00  사업개시   
265   서귀포시  남원읍  2014-05-15  (유)무한에코에너지(서민행복8호태양광발전소)    516.00  사업개시   
417   서귀포시  강정동  2014-12-09             (유)용흥마을태양광발전소     99.00  사업개시   
343   서귀포시  성산읍  2014-10-14                 (유)제주 늘해랑   2198.00  사업개시   
...    ...  ...         ...                       ...       ...   ...   
1069  서귀포시  표선면  2018-09-19                후니3 태양광발전소     93.15  사업개시   
353   서귀포시  성산읍  2014-10-28                희윤2 태양광발전소     99.00  사업개시   
357   서귀포시  성산읍  2014-10-29                희윤4 태양광발전소     99.28  사업개시   
359   서귀포시  성산읍  2014-10-29                희윤5 태양광발전소     99.28  사업개시   
358   서귀포시  성산읍  2014-10-29                  희윤태양광발전소     99.28  사업개시   

           사업개시일     데이터기준일자  
804   2019-04-26  2022-06-30  
329   2015-07-01  2022-06-30  
265   2014-10-16  2022-06-30  
417   2015-10-29  2022-06-30  
343   2019-01-31  2022-06-30  
...          ...         ...  
1069  2019-03-04  2022-06-30  
353   2015-06-22  2022-06-30  
357   2019-07-11  2022-06-30  
359   2019-07-11  2022-06-30  
358   2019-07-11  2022-06-30  

[770 rows x 8 columns]

In [32]:
solar_df = df.loc[indexList]

## 이제 지도에 표시

In [33]:
solar_df['location'] = solar_df['행정시'] + ' ' + solar_df['읍면동']
solar_df.head()

행정시  읍면동        허가일자                        상호  설비용량(KW)    상태  \
804   제주시  애월읍  2018-04-04           (유)나월(나월태양광발전소)     496.8  사업개시   
329   제주시  구좌읍  2014-10-07            (유)메가솔라 태양광발전소    1800.0  사업개시   
265  서귀포시  남원읍  2014-05-15  (유)무한에코에너지(서민행복8호태양광발전소)     516.0  사업개시   
417  서귀포시  강정동  2014-12-09             (유)용흥마을태양광발전소      99.0  사업개시   
343  서귀포시  성산읍  2014-10-14                 (유)제주 늘해랑    2198.0  사업개시   

          사업개시일     데이터기준일자  location  
804  2019-04-26  2022-06-30   제주시 애월읍  
329  2015-07-01  2022-06-30   제주시 구좌읍  
265  2014-10-16  2022-06-30  서귀포시 남원읍  
417  2015-10-29  2022-06-30  서귀포시 강정동  
343  2019-01-31  2022-06-30  서귀포시 성산읍

In [34]:
solar_df.reset_index(drop = True, inplace = True)

In [35]:
from geopy.geocoders import Nominatim

def geocoding(address):
    geolocoder = Nominatim(user_agent = 'South Korea', timeout=None)
    geo = geolocoder.geocode(address)
    if geo==None:
        print('geo is null')
        return 0, 0
    else:
        return geo.latitude, geo.longitude


geocoding(solar_df.loc[0, 'location'])

(33.45079, 126.3752)

In [36]:
a, b = geocoding(solar_df.loc[0, 'location'])
a

33.45079

In [37]:
# latitude = []
# longitude = []

# for i in range(solar_df.shape[0]):
#   a, b = geocoding(solar_df.loc[i, 'location'])
#   latitude.append(a)
#   longitude.append(b)

# solar_df['latitude'] = latitude
# solar_df['longitude'] = longitude

# solar_df.head()

이렇게 770개 다 돌리면 시간이 오래걸림. 그래서 그 중 유니크한 값들로 추려서 지오코딩 변환할 것.

In [38]:
list(solar_df['location'].unique())

array(['제주시 애월읍', '제주시 구좌읍', '서귀포시 남원읍', '서귀포시 강정동', '서귀포시 성산읍',
       '제주시 한경면', '제주시 회천동', '서귀포시 대정읍', '서귀포시 표선면', '제주시 영평동',
       '서귀포시 하원동', '제주시 노형동', '제주시 한림읍', '서귀포시 중문동', '서귀포시 동홍동',
       '서귀포시 상효동', '서귀포시 서홍동', '서귀포시 안덕면', '제주시 용담이동', '제주시 조천읍',
       '서귀포시 상예동', '제주시 삼도이동', '제주시 도련일동', '서귀포시 도순동', '제주시 오등동',
       '제주시 봉개동', '제주시 외도일동', '제주시 해안동', '서귀포시 하효동', '제주시 화북이동',
       '제주시 이호이동', '제주시 오라이동', '서귀포시 서귀동', '제주시 용강동', '서귀포시 서호동',
       '제주시 삼양일동', '서귀포시 토평동', '서귀포시 호근동', '제주시 도련이동', '제주시 아라이동',
       '제주시 아라일동', '제주시 오라일동', '서귀포시 하예동', '제주시 이도일동', '서귀포시 색달동',
       '제주시 도평동'], dtype=object)

In [40]:
unique_addr=[]
unique_lat=[]
unique_lng=[]

for addr in solar_df['location'].unique():
    print(addr)
    lat, lng = geocoding(addr)
    if lat==0 or lng==0:
        continue
    unique_addr.append(addr)
    unique_lat.append(lat)
    unique_lng.append(lng)
    print(lat, lng)

제주시 애월읍
33.45079 126.3752
제주시 구좌읍
33.51662 126.81055
서귀포시 남원읍
33.3076683 126.6907475
서귀포시 강정동
33.24695 126.48927
서귀포시 성산읍
33.41805 126.88147
제주시 한경면
33.3305 126.21066
제주시 회천동
33.4956 126.61479
서귀포시 대정읍
33.2218157 126.2529945
서귀포시 표선면
33.3504543 126.794049
제주시 영평동
33.4812 126.57106
서귀포시 하원동
33.2626587 126.4600239
제주시 노형동
33.4781247 126.4754867
제주시 한림읍
33.39344 126.26684
서귀포시 중문동
33.262002 126.4288617
서귀포시 동홍동
33.26184 126.56804
서귀포시 상효동
33.2896254 126.5945088
서귀포시 서홍동
33.25965 126.55175
서귀포시 안덕면
33.27688 126.34925
제주시 용담이동
33.50742 126.50615
제주시 조천읍
33.50888 126.66305
서귀포시 상예동
33.26177 126.38678
제주시 삼도이동
geo is null
제주시 도련일동
33.5059 126.5869
서귀포시 도순동
33.2569662 126.4731036
제주시 오등동
33.46148 126.53087
제주시 봉개동
33.484188 126.5993497
제주시 외도일동
33.4855847 126.4284514
제주시 해안동
33.44899 126.46036
서귀포시 하효동
33.2582539 126.6179833
제주시 화북이동
33.50469 126.5648
제주시 이호이동
33.498551 126.4565213
제주시 오라이동
33.47081 126.50991
서귀포시 서귀동
33.24679 126.56396
제주시 용강동
33.47017 126.58934
서귀포시 서호동
33.25759 126.51916
제주

In [44]:
geocoding('제주시 삼도2동')

geo is null


(0, 0)

[주소 영문 변환](https://jusoga.com/%EC%A0%9C%EC%A3%BC%ED%8A%B9%EB%B3%84%EC%9E%90%EC%B9%98%EB%8F%84/%EC%A0%9C%EC%A3%BC%EC%8B%9C/%EC%82%BC%EB%8F%84%EC%9D%B4%EB%8F%99)

In [45]:
geocoding('Gwandeok-ro, Jeju-si')

(33.5128197, 126.5199591)

이걸 하나의 데이터프레임으로 합치면 됨.